org.eclipse.scanning.api/src/org/eclipse/scanning/api/device/DeviceRequestHandler.java

```diff
@@ -1,249 +1,249 @@
+			error.merge(request);
+			error.setErrorMessage(ne.getMessage());
+			error.setErrorFieldNames(ne.getFieldNames());
+			return error;
+
+		} catch (Exception ne) {
+			ne.printStackTrace();
+			DeviceRequest error = new DeviceRequest();
+			error.merge(request);
+			error.setErrorMessage(ne.getMessage());
+			return error;
+		}
+	}
+
+	private static void processScannables(DeviceRequest request, IScannableDeviceService cservice) throws Exception {
+
+		if (request.getDeviceName()!=null) { // Named device required
+
+			IScannable<Object> device = cservice.getScannable(request.getDeviceName());
+			DeviceAction action = request.getDeviceAction();
+			if (action==DeviceAction.SET && request.getDeviceValue()!=null) {
+
+				// Not sure if using type is a good design idea.
+				if (request.getDeviceValue() instanceof MonitorRole) {
+					device.setMonitorRole((MonitorRole)request.getDeviceValue());
+				} else {
+					device.setPosition(request.getDeviceValue(), request.getPosition());
+					/* This thread is executing to set position, while it does that it
+					 * sends events over AMQ. These events queue up with no higher priority
+					 * than this call. Therefore if we return immediately it is possible
+					 * for this call to return before position events are sent out.
+					 * FUDGE warning
+					 */
+					Thread.sleep(10);
+					 /* End warning */
+				}
+			} else if (action==DeviceAction.ACTIVATE) {
+				device.setActivated((Boolean)request.getDeviceValue());
+			}
+
+
+			if (action!=null && action.isTerminate() && device instanceof ITerminatable) {
+				ITerminatable tdevice = (ITerminatable)device;
+				tdevice.terminate(action.to());
+			}
+
+			request.setDeviceValue(device.getPosition());
+
+			DeviceInformation<?> info = new DeviceInformation<Object>(device.getName());
+			merge(info, device);
+			request.addDeviceInformation(info);
+
+		} else {
+			final Collection<String> names = cservice.getScannableNames();
+			for (String name : names) {
+
+				if (name==null) continue;
+				if (request.getDeviceName()!=null && !name.matches(request.getDeviceName())) continue;
+
+				IScannable<?> device = cservice.getScannable(name);
+				if (device==null) throw new EventException("There is no created device called '"+name+"'");
+
+				DeviceInformation<?> info = new DeviceInformation<Object>(name);
+				merge(info, device);
+				request.addDeviceInformation(info);
+			}
+		}
+	}
+
+
+	private static void merge(DeviceInformation<?> info, IScannable<?> device) throws Exception {
+		info.setLevel(device.getLevel());
+		info.setUnit(device.getUnit());
+        info.setUpper(device.getMaximum());
+        info.setLower(device.getMinimum());
+        info.setPermittedValues(device.getPermittedValues());
+        info.setActivated(device.isActivated());
+        info.setMonitorRole(device.getMonitorRole());
+	}
+
+	private static void processRunnables(DeviceRequest request, IRunnableDeviceService dservice) throws Exception {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 32 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)